HLTrigger/HLTfilters/plugins/HLTDoubletDZ.cc

```diff
@@ -310,348 +342,271 @@ HLTDoubletDZ<T1, T2>::getCollections(edm::Event& iEvent,
 }
 
-template <>
-bool HLTDoubletDZ<l1t::HPSPFTau, l1t::HPSPFTau>::computeDZ(edm::Event& iEvent,
-                                                           l1t::HPSPFTauRef& r1,
-                                                           l1t::HPSPFTauRef& r2) const {
-  const l1t::HPSPFTau& candidate1(*r1);
-  const l1t::HPSPFTau& candidate2(*r2);
-  if (reco::deltaR(candidate1, candidate2) < minDR_)
-    return false;
-
-  // We don't care about minPixHitsForDZ_ with the L1HPSPFTaus,
-  // especially because the pixel does not participate in the L1T
-  if (std::abs(candidate1.leadChargedPFCand()->pfTrack()->vertex().z() -
-               candidate2.leadChargedPFCand()->pfTrack()->vertex().z()) > maxDZ_)
-    return false;
+namespace {
 
-  return true;
-}
+  double getCandidateVZ(reco::RecoChargedCandidate const& cand, bool& isValidVZ, int const minPixHitsForValidVZ) {
+    if (minPixHitsForValidVZ > 0) {
+      auto const track = cand.track();
+      if (not(track.isNonnull() and track.isAvailable() and
+              track->hitPattern().numberOfValidPixelHits() >= minPixHitsForValidVZ)) {
+        isValidVZ = false;
+        return 0;
+      }
+    }
 
-// ------------ method called to produce the data  ------------
-template <typename T1, typename T2>
-bool HLTDoubletDZ<T1, T2>::hltFilter(edm::Event& iEvent,
-                                     const edm::EventSetup& iSetup,
-                                     trigger::TriggerFilterObjectWithRefs& filterproduct) const {
-  // All HLT filters must create and fill an HLT filter object,
-  // recording any reconstructed physics objects satisfying (or not)
-  // this HLT filter, and place it in the Event.
-  bool accept(false);
+    isValidVZ = true;
+    return cand.vz();
+  }
 
-  std::vector<T1Ref> coll1;
-  std::vector<T2Ref> coll2;
+  double getCandidateVZ(reco::RecoEcalCandidate const& cand,
+                        bool& isValidVZ,
+                        int const minPixHitsForValidVZ,
+                        reco::ElectronCollection const& electrons) {
+    reco::Electron const* elec_ptr = nullptr;
+    for (auto const& elec : electrons) {
+      if (elec.superCluster() == cand.superCluster()) {
+        elec_ptr = &elec;
+      }
+    }
 
-  if (getCollections(iEvent, coll1, coll2, filterproduct)) {
-    int n(0);
-    T1Ref r1;
-    T2Ref r2;
+    if (elec_ptr == nullptr) {
+      // IMPROVE, kept for backward compatibility
+      isValidVZ = true;
+      reco::Electron e1;
+      return e1.vz();
+    }
 
-    for (unsigned int i1 = 0; i1 != coll1.size(); i1++) {
-      r1 = coll1[i1];
-      //const reco::Candidate& candidate1(*r1);
-      unsigned int I(0);
-      if (same_) {
-        I = i1 + 1;
+    if (minPixHitsForValidVZ > 0) {
+      auto const track = elec_ptr->gsfTrack();
+      if (not(track.isNonnull() and track.isAvailable() and
+              track->hitPattern().numberOfValidPixelHits() >= minPixHitsForValidVZ)) {
```

Comment:
This is ultimately the change that addresses #40615. A similar check is also added for other types that attempt to access tracks via Refs (e.g. `reco::RecoChargedCandidate`).